# Integration Testing, Coverage & CI/CD
** Integration Testing | Coverage |  CI/CD**

---

### Learning Objectives
- The difference between unit, integration, and E2E tests
- `TestClient` from FastAPI — test HTTP endpoints without running a server
- Testing GET, POST, PUT, DELETE — status codes, response body, headers
- Testing error cases (404, 422 validation errors, 400 bad requests)
- Dependency Override — swap out dependencies for testing
- Strategy for testing with a real (but isolated) database
- Test coverage with `pytest-cov` — what it measures and why 80%+ matters
- Reading coverage reports (terminal, HTML)
- CI/CD: what it is and why it exists
- GitHub Actions anatomy: workflows, jobs, steps, runners
- Writing a real CI workflow YAML

---
> **Note:** This notebook has conceptual runnable cells AND reference examples.
> The `examples/` folder in this repo has a complete FastAPI app + full test suite to run with `pytest` from terminal.

## 1. Unit vs Integration vs E2E — The Real Difference

```
            What is being tested          Real dependencies?
Unit        One function in isolation     None (all mocked)
Integration Multiple components together  Some real (DB), some mocked (external APIs)
E2E         Full user flow               Everything real (browser, DB, network)
```

### Unit Test Example:
```python
def test_hash_password():                 # tests ONE function
    hashed = hash_password('secret123')
    assert len(hashed) == 64             # no DB, no HTTP, just logic
```

### Integration Test Example:
```python
def test_create_user_endpoint(client, test_db):   # tests API + DB together
    response = client.post('/users', json={'name': 'Alice', 'email': 'a@b.com'})
    assert response.status_code == 201
    assert test_db.count('users') == 1    # user actually saved to real test DB
```

### Why this matters for FastAPI:
Integration tests for FastAPI endpoints should use **`TestClient`** — it sends real HTTP requests through the full ASGI stack (middleware, routing, validation) but without starting an actual server on a port. This means you test the full request/response cycle at HTTP level, including Pydantic validation.

## 2. FastAPI `TestClient` — How It Works

```python
from fastapi.testclient import TestClient
from app.main import app          # your FastAPI app

client = TestClient(app)          # wraps the ASGI app in a sync HTTP client
```

**Under the hood:** `TestClient` uses `httpx` (with an ASGI transport) to send requests directly into your FastAPI app in-process. No port, no network, no server process.

**What it tests:**
- Route matching
- Pydantic request body validation
- Response serialization
- Middleware execution
- Dependency injection
- HTTP status codes
- Authentication & authorization

**What it doesn't test:**
- Actual network latency
- Load/performance
- Multi-process state

In [1]:
%%writefile mini_app.py
# mini_app.py — minimal FastAPI app for TestClient testing demos
# uses an in-memory dict instead of a real DB — keeps the focus on testing patterns
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from typing import Optional

app = FastAPI()

# ─── In-memory store ──────────────────────────────────────────────────────────
# simulates a database using a plain dict — resets every time the app restarts
# _next_id auto-increments to simulate a DB sequence/serial primary key
_users: dict = {}
_next_id: int = 1


# ─── Pydantic models ──────────────────────────────────────────────────────────
# UserCreate  → what the CLIENT sends in the request body
# UserResponse → what the SERVER sends back — includes the auto-generated id

class UserCreate(BaseModel):
    name: str
    email: str
    age: Optional[int] = None       # optional field — can be omitted from request


class UserResponse(BaseModel):
    id: int                          # added by the server, not provided by client
    name: str
    email: str
    age: Optional[int] = None


# ─── Routes ───────────────────────────────────────────────────────────────────

@app.get('/health')
def health_check():
    # simple liveness check — useful in tests to verify the app is running
    return {'status': 'ok'}


@app.post('/users', response_model=UserResponse, status_code=201)
def create_user(user: UserCreate):
    global _next_id                             # modify the module-level counter
    new_user = {'id': _next_id, **user.model_dump()}  # merge id + all user fields
    _users[_next_id] = new_user                 # store in dict with id as key
    _next_id += 1                               # increment for next user
    return new_user                             # FastAPI validates against UserResponse


@app.get('/users', response_model=list[UserResponse])
def list_users():
    # return all users as a list — empty list if no users exist yet
    return list(_users.values())


@app.get('/users/{user_id}', response_model=UserResponse)
def get_user(user_id: int):
    if user_id not in _users:
        raise HTTPException(status_code=404, detail=f'User {user_id} not found')
    return _users[user_id]


@app.put('/users/{user_id}', response_model=UserResponse)
def update_user(user_id: int, user: UserCreate):
    if user_id not in _users:
        raise HTTPException(status_code=404, detail=f'User {user_id} not found')
    _users[user_id].update(user.model_dump())   # overwrite existing fields with new values
    return _users[user_id]


@app.delete('/users/{user_id}', status_code=204)
def delete_user(user_id: int):
    if user_id not in _users:
        raise HTTPException(status_code=404, detail=f'User {user_id} not found')
    del _users[user_id]
    # 204 No Content — no return value, response body is empty by convention

Writing mini_app.py


## 3. Testing GET Endpoints

In [2]:
%%writefile test_mini_get.py
import pytest
from fastapi.testclient import TestClient
from mini_app import app, _users, _next_id


# ─── State reset fixture ──────────────────────────────────────────────────────
# autouse=True means this fixture runs automatically for EVERY test in this file
# without autouse, leftover state from one test would bleed into the next
# e.g. a user created in test_1 would still exist when test_2 runs

@pytest.fixture(autouse=True)
def clear_state():
    """Reset in-memory store before each test."""
    import mini_app
    mini_app._users.clear()      # wipe all users — start from empty dict
    mini_app._next_id = 1        # reset ID counter — IDs are predictable again
    yield                         # test runs here
                                  # no teardown needed — next test resets again anyway


# ─── TestClient fixture ───────────────────────────────────────────────────────
# TestClient wraps the FastAPI app and lets you make HTTP requests in tests
# without spinning up a real server — requests are handled in-process

@pytest.fixture
def client():
    return TestClient(app)


# ─── Health check ─────────────────────────────────────────────────────────────

def test_health_check(client):
    response = client.get('/health')
    assert response.status_code == 200
    assert response.json() == {'status': 'ok'}     # exact body match


# ─── GET /users — empty state ─────────────────────────────────────────────────
# tests the empty case first — always test the base case before adding data

def test_list_users_empty(client):
    response = client.get('/users')
    assert response.status_code == 200
    assert response.json() == []                    # empty list, not null or 404


# ─── GET /users/{id} — existing user ─────────────────────────────────────────
# arrange: create a user first (POST)
# act:     fetch it by id (GET)
# assert:  verify the response matches what was created

def test_get_user_by_id(client):
    # arrange — create a user so there's something to fetch
    client.post('/users', json={'name': 'Alice', 'email': 'alice@example.com'})

    # act — fetch by id (id=1 because _next_id was reset to 1 by clear_state)
    response = client.get('/users/1')

    # assert — verify status and each field individually
    assert response.status_code == 200
    data = response.json()
    assert data['id'] == 1
    assert data['name'] == 'Alice'
    assert data['email'] == 'alice@example.com'


# ─── GET /users/{id} — non-existent user ─────────────────────────────────────
# always test the unhappy path — what happens when things go wrong

def test_get_user_not_found(client):
    response = client.get('/users/999')             # no user with id 999
    assert response.status_code == 404
    assert 'not found' in response.json()['detail'].lower()   # check error message content
                                                               # .lower() makes it case-insensitive


# ─── GET /users — with multiple users ────────────────────────────────────────
# tests that list endpoint returns ALL users in insertion order

def test_list_users_with_data(client):
    # arrange — seed two users
    client.post('/users', json={'name': 'Alice', 'email': 'alice@example.com'})
    client.post('/users', json={'name': 'Bob',   'email': 'bob@example.com'})

    # act
    response = client.get('/users')

    # assert — check count AND order AND content
    assert response.status_code == 200
    users = response.json()
    assert len(users) == 2                  # exactly 2 users, no more no less
    assert users[0]['name'] == 'Alice'      # insertion order preserved
    assert users[1]['name'] == 'Bob'

Writing test_mini_get.py


In [3]:
!pytest test_mini_get.py -v

============================= test session starts ==============================
platform darwin -- Python 3.11.14, pytest-9.1.1, pluggy-1.6.0 -- /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/myenv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/backend-and-systems/testing/notebooks
plugins: cov-7.1.0, anyio-4.13.0
collected 5 items                                                              

test_mini_get.py::test_health_check PASSED                               [ 20%]
test_mini_get.py::test_list_users_empty PASSED                           [ 40%]
test_mini_get.py::test_get_user_by_id PASSED                             [ 60%]
test_mini_get.py::test_get_user_not_found PASSED                         [ 80%]
test_mini_get.py::test_list_users_with_data PASSED                       [100%]

=============================== warnings summary ===============================
../../../myenv/lib/python3

## 4. Testing POST, PUT, DELETE — and Validation Errors

In [4]:
%%writefile test_mini_write.py
import pytest
from fastapi.testclient import TestClient
from mini_app import app


@pytest.fixture(autouse=True)
def clear_state():
    import mini_app
    mini_app._users.clear()
    mini_app._next_id = 1
    yield


@pytest.fixture
def client():
    return TestClient(app)


# --- POST: create user ---
def test_create_user(client):
    response = client.post('/users', json={
        'name': 'Alice',
        'email': 'alice@example.com',
        'age': 30
    })
    assert response.status_code == 201              # 201 Created
    data = response.json()
    assert data['id'] == 1
    assert data['name'] == 'Alice'
    assert data['age'] == 30


def test_create_user_minimal_fields(client):
    """age is optional — should still work without it."""
    response = client.post('/users', json={'name': 'Bob', 'email': 'bob@example.com'})
    assert response.status_code == 201
    assert response.json()['age'] is None


# --- POST: validation errors (422 Unprocessable Entity) ---
def test_create_user_missing_required_field(client):
    """'email' is required — omitting it triggers Pydantic validation."""
    response = client.post('/users', json={'name': 'Alice'})   # no email
    assert response.status_code == 422                          # FastAPI validation error
    errors = response.json()['detail']
    assert any('email' in str(e).lower() for e in errors)


def test_create_user_empty_body(client):
    response = client.post('/users', json={})
    assert response.status_code == 422


# --- PUT: update user ---
def test_update_user(client):
    client.post('/users', json={'name': 'Alice', 'email': 'old@example.com'})

    response = client.put('/users/1', json={
        'name': 'Alice Updated',
        'email': 'new@example.com'
    })
    assert response.status_code == 200
    assert response.json()['name'] == 'Alice Updated'
    assert response.json()['email'] == 'new@example.com'


def test_update_nonexistent_user(client):
    response = client.put('/users/999', json={'name': 'Ghost', 'email': 'g@example.com'})
    assert response.status_code == 404


# --- DELETE ---
def test_delete_user(client):
    client.post('/users', json={'name': 'Alice', 'email': 'alice@example.com'})

    response = client.delete('/users/1')
    assert response.status_code == 204               # 204 No Content

    # Verify user is gone
    get_response = client.get('/users/1')
    assert get_response.status_code == 404


def test_delete_nonexistent_user(client):
    response = client.delete('/users/999')
    assert response.status_code == 404


# --- Full CRUD flow ---
def test_full_crud_flow(client):
    """Tests create → read → update → delete as one flow."""
    # Create
    res = client.post('/users', json={'name': 'Alice', 'email': 'alice@example.com'})
    assert res.status_code == 201
    user_id = res.json()['id']

    # Read
    res = client.get(f'/users/{user_id}')
    assert res.json()['name'] == 'Alice'

    # Update
    res = client.put(f'/users/{user_id}', json={'name': 'Alice New', 'email': 'alice@example.com'})
    assert res.json()['name'] == 'Alice New'

    # Delete
    res = client.delete(f'/users/{user_id}')
    assert res.status_code == 204

    # Verify gone
    assert client.get(f'/users/{user_id}').status_code == 404

Writing test_mini_write.py


In [5]:
!pytest test_mini_write.py -v

============================= test session starts ==============================
platform darwin -- Python 3.11.14, pytest-9.1.1, pluggy-1.6.0 -- /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/myenv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/backend-and-systems/testing/notebooks
plugins: cov-7.1.0, anyio-4.13.0
collected 9 items                                                              

test_mini_write.py::test_create_user PASSED                              [ 11%]
test_mini_write.py::test_create_user_minimal_fields PASSED               [ 22%]
test_mini_write.py::test_create_user_missing_required_field PASSED       [ 33%]
test_mini_write.py::test_create_user_empty_body PASSED                   [ 44%]
test_mini_write.py::test_update_user PASSED                              [ 55%]
test_mini_write.py::test_update_nonexistent_user PASSED                  [ 66%]
test_mini_write.py::test_del

## 5. Dependency Override — Swap Dependencies in Tests

FastAPI lets you **replace any dependency with a different one at test
time** — without touching the actual app code. Same idea as mocking,
but built into FastAPI's DI system.

---

### The Problem Without It

```python
# app code
def get_db():
    return ProductionDatabase()    # connects to real DB

@app.get('/users')
def list_users(db = Depends(get_db)):
    return db.get_all_users()
```

```python
# test — without override
def test_list_users(client):
    response = client.get('/users')   # ❌ hits the real production DB
                                       # slow, state-dependent, dangerous
```

---

### The Solution — `dependency_overrides`

```python
# test version of the dependency
def override_get_db():
    db = TestDatabase()            # in-memory, isolated, fast
    yield db
    db.close()

# swap it in — key: real function, value: replacement function
app.dependency_overrides[get_db] = override_get_db

client = TestClient(app)
# now every endpoint that uses get_db gets TestDatabase instead
# app code is completely unchanged
```

---

### Always Clean Up — Prevent Test Pollution

If you don't clear overrides after a test, every test that runs after
it also gets the override — even ones that don't want it.

```python
# ❌ without cleanup — override leaks into other tests
def test_list_users():
    app.dependency_overrides[get_db] = override_get_db
    ...
    # override is still active after this test finishes

# ✅ with cleanup — use a fixture so teardown is guaranteed
@pytest.fixture
def client():
    app.dependency_overrides[get_db] = override_get_db   # swap before test
    yield TestClient(app)                                  # test runs here
    app.dependency_overrides.clear()                       # restore after test
                                                           # runs even if test fails
```

---

### What You Can Override

```python
# swap the DB — use test DB instead of production DB
app.dependency_overrides[get_db] = override_get_db

# bypass auth — test as a specific user without a real token
app.dependency_overrides[get_current_user] = lambda: {'id': 1, 'role': 'admin'}

# replace external API client — no real HTTP calls in tests
app.dependency_overrides[get_weather_client] = lambda: MockWeatherClient()
```

Any dependency in `Depends()` can be overridden — no changes to app
code required.

---

### How It Fits Together

```
Normal request:  endpoint → Depends(get_db)          → ProductionDatabase
Test request:    endpoint → Depends(get_db) overridden → TestDatabase
                           ↑
                app.dependency_overrides handles this swap transparently
```

---

> 💡 `dependency_overrides` is a plain dict on the `app` object —
> key is the real dependency function, value is the replacement.
> `clear()` removes all overrides at once, restoring normal behavior.

In [6]:
%%writefile test_dependency_override.py
import pytest
from fastapi import FastAPI, Depends
from fastapi.testclient import TestClient


# --- App with a dependency ---
app = FastAPI()

def get_settings():
    """Returns app config. In production: reads from env vars."""
    return {'max_users': 1000, 'environment': 'production', 'debug': False}

def get_current_user():
    """In production: reads JWT, validates token, returns user."""
    return {'id': 1, 'name': 'Real User', 'role': 'user'}


@app.get('/config')
def read_config(settings: dict = Depends(get_settings)):
    return settings


@app.get('/profile')
def read_profile(user: dict = Depends(get_current_user)):
    return {'user': user, 'message': f'Hello {user["name"]}'}


@app.get('/admin')
def admin_panel(user: dict = Depends(get_current_user)):
    if user['role'] != 'admin':
        from fastapi import HTTPException
        raise HTTPException(status_code=403, detail='Admins only')
    return {'message': 'Welcome to admin panel'}


# --- Tests using dependency overrides ---

@pytest.fixture
def client():
    """Client with test settings override."""
    def test_settings():
        return {'max_users': 10, 'environment': 'test', 'debug': True}

    app.dependency_overrides[get_settings] = test_settings
    yield TestClient(app)
    app.dependency_overrides.clear()      # IMPORTANT: restore after test


@pytest.fixture
def admin_client():
    """Client authenticated as admin."""
    def fake_admin():
        return {'id': 99, 'name': 'Test Admin', 'role': 'admin'}

    app.dependency_overrides[get_current_user] = fake_admin
    yield TestClient(app)
    app.dependency_overrides.clear()


def test_config_override(client):
    response = client.get('/config')
    assert response.status_code == 200
    data = response.json()
    assert data['environment'] == 'test'    # overridden!
    assert data['debug'] is True


def test_profile_as_regular_user(client):
    response = client.get('/profile')
    # get_current_user NOT overridden here — uses real dependency
    assert response.status_code == 200


def test_admin_access_granted(admin_client):
    response = admin_client.get('/admin')
    assert response.status_code == 200
    assert response.json()['message'] == 'Welcome to admin panel'


def test_admin_access_denied_for_regular_user(client):
    """Regular user fixture — get_current_user returns role='user'."""
    response = client.get('/admin')
    assert response.status_code == 403
    assert response.json()['detail'] == 'Admins only'

Writing test_dependency_override.py


In [7]:
!pytest test_dependency_override.py -v

============================= test session starts ==============================
platform darwin -- Python 3.11.14, pytest-9.1.1, pluggy-1.6.0 -- /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/myenv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/anishupreti2002icloud.com/Desktop/Learning/relearning-concepts/backend-and-systems/testing/notebooks
plugins: cov-7.1.0, anyio-4.13.0
collected 4 items                                                              

test_dependency_override.py::test_config_override PASSED                 [ 25%]
test_dependency_override.py::test_profile_as_regular_user PASSED         [ 50%]
test_dependency_override.py::test_admin_access_granted PASSED            [ 75%]
test_dependency_override.py::test_admin_access_denied_for_regular_user PASSED [100%]

=============================== warnings summary ===============================
../../../myenv/lib/python3.11/site-packages/fastapi/testclient.py:1
  /Users/anishupreti2002icloud.co

## 6. Database Testing Strategy

```
Production DB  →  PostgreSQL  (real data, never touch in tests)
Test DB        →  SQLite      (fast, in-memory, disposable)
```

The test DB uses the **same SQLAlchemy models** as production — only
the connection URL changes. Tests never know the difference.

---

### App Side — Production DB Setup

```python
# database.py — production setup, never modified for tests
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

DATABASE_URL = 'postgresql://user:pass@localhost/mydb'
engine = create_engine(DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def get_db():
    db = SessionLocal()
    try:
        yield db          # FastAPI injects this session into endpoints
    finally:
        db.close()
```

---

### Test Side — SQLite Override in `conftest.py`

```python
# conftest.py — three fixtures that chain into each other
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from app.database import Base, get_db
from app.main import app

TEST_DATABASE_URL = 'sqlite:///./test.db'   # or 'sqlite:///:memory:' for pure in-memory


@pytest.fixture(scope='module')
def test_engine():
    # scope='module' — tables created ONCE for the whole file, not per test
    engine = create_engine(TEST_DATABASE_URL, connect_args={'check_same_thread': False})
    Base.metadata.create_all(bind=engine)    # create all tables from your models
    yield engine
    Base.metadata.drop_all(bind=engine)      # drop all tables after the module finishes


@pytest.fixture
def db_session(test_engine):
    TestingSessionLocal = sessionmaker(bind=test_engine)
    session = TestingSessionLocal()
    yield session                # test runs here — uses this session
    session.rollback()           # ← undo ALL changes from this test in one shot
    session.close()


@pytest.fixture
def client(db_session):
    def override_get_db():
        yield db_session          # hand the TEST session to every endpoint

    app.dependency_overrides[get_db] = override_get_db   # swap production DB → test DB
    yield TestClient(app)
    app.dependency_overrides.clear()                       # restore after test
```

---

### Fixture Chain — How They Connect

```
test_engine (module scope)     — creates tables once
    └── db_session (function)  — opens a session per test, rolls back after
            └── client         — swaps get_db → test session, gives you TestClient
```

Each test just requests `client` — the whole chain resolves automatically.

---

### Why `rollback()` Instead of Deleting Records?

```
session.rollback()   →  undoes ALL changes instantly — one operation
deleting records     →  must know what was inserted, risk of missing something
```

Rollback leaves the DB in exactly the same state it was before the
test — no partial cleanup, no order dependency between tests.

---

> 💡 The three fixtures always go together as a set. Copy this pattern
> into every project's `conftest.py` and you get isolated, fast, safe
> DB testing out of the box.

## 7. Test Coverage — What Is It?

Coverage answers one question: **"which lines of your code actually
ran when your tests executed?"** Lines that never ran = bugs hiding
in untested branches.

---

### The Problem — Untested Branches

```python
def process_order(order):
    if order['status'] == 'pending':          # ✅ tested
        return send_email(order['user'])      # ✅ tested
    elif order['status'] == 'cancelled':      # ❌ never reached
        return refund(order['payment_id'])    # ❌ never reached
    else:
        raise ValueError('Unknown status')   # ❌ never reached
```

If your tests only call this with `status='pending'`, the bottom two
branches never run. A bug in `refund()` logic would ship undetected.
Coverage makes this **visible** — it flags lines 3, 4, 5 as uncovered.

---

### Types of Coverage

| Type | What it checks | Strictness |
|------|---------------|------------|
| **Line** | Was this line executed? | Default — most common |
| **Branch** | Was every `if/else` path taken? | Stricter — catches missing `else` |
| **Function** | Was this function called at all? | Coarsest — least useful alone |

Branch coverage is stricter than line coverage:

```python
if user.is_admin:
    grant_access()
# no else — line coverage says ✅ (the if line ran)
# branch coverage says ❌ (the False branch was never taken)
```

---

### What 80%+ Actually Means

```
80% is a common team minimum — not a target to aim for and stop at

✅ critical paths (payment, auth, core logic) → aim for 100%
✅ error handlers and edge cases              → must be tested
⚠️  logging helpers, config loaders           → 60-70% is acceptable
❌ hitting 80% by testing only the happy path → false confidence
```

---

### The Most Important Caveat

```python
def add(a, b):
    return a - b       # bug: should be a + b

def test_add():
    result = add(1, 1)
    assert result is not None    # ✅ line was executed → 100% coverage
                                  # ❌ wrong assertion — bug goes unnoticed
```

**Coverage tells you what ran — not whether it was tested correctly.**
100% coverage with weak assertions is still broken code.

---

> 💡 Coverage is a tool to find gaps — lines you forgot to test at
> all. It's not a quality score. Low coverage = definite problem.
> High coverage = necessary but not sufficient.

## 8. `pytest-cov` — Running Coverage

```bash
# Install
pip install pytest-cov

# Run tests with coverage for a specific source
pytest --cov=app tests/

# Show which lines are missing
pytest --cov=app --cov-report=term-missing tests/

# Generate HTML report (open htmlcov/index.html)
pytest --cov=app --cov-report=html tests/

# Branch coverage (stricter)
pytest --cov=app --cov-branch tests/

# Fail if coverage is below 80%
pytest --cov=app --cov-fail-under=80 tests/
```

### Terminal output with `--cov-report=term-missing`:
```
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
app/main.py              42      8    81%   23-25, 67, 71-75
app/models.py            18      0   100%
app/utils.py             31     12    61%   45-52, 58-63
---------------------------------------------------
TOTAL                    91     20    78%
```

**Reading this:**
- `Stmts`: total executable lines
- `Miss`: lines NOT executed during tests
- `Cover`: `(Stmts - Miss) / Stmts * 100`
- `Missing`: specific line numbers you need to add tests for

## 9. `.coveragerc` — Coverage Configuration

`.coveragerc` lives in your project root and tells coverage.py exactly
what to measure, what to ignore, and how to format the report. Without
it, coverage measures everything including files you don't care about.

```
project-root/
  .coveragerc        ← coverage reads this automatically when you run pytest --cov
  app/
  tests/
  requirements.txt
```

---

```ini
# .coveragerc

# ── [run] controls WHAT gets measured ────────────────────────────────────────
[run]

source = app
# Only measure coverage inside the 'app/' package.
# Without this, coverage also measures test files, venv, stdlib — noise you don't want.
# Equivalent to: pytest --cov=app (but set once here instead of every command)

branch = true
# Enable BRANCH coverage — stricter than line coverage.
# Line coverage: was this line executed? (yes/no)
# Branch coverage: was EVERY path through this line taken?
#
# Example where branch coverage catches what line coverage misses:
#   if user.is_admin:     ← line coverage says: executed = covered ✅
#       delete_all()      ← branch coverage asks: was the ELSE path (non-admin) also tested?

omit =
    app/__init__.py       # usually empty or just imports — not worth measuring
    app/config.py         # reads env vars — behaves differently in test vs prod,
                          # excluding keeps your coverage number honest
    */migrations/*        # auto-generated DB migration files — not your logic


# ── [report] controls WHAT appears in the report ─────────────────────────────
[report]

exclude_lines =
# these LINE PATTERNS are excluded from coverage even if never executed
# coverage.py checks each line against these regex patterns

    pragma: no cover      # any line with this comment is excluded — added manually
                          # use sparingly — only for lines genuinely impossible to test

    def __repr__          # debugging helper — not usually called in tests

    raise NotImplementedError   # abstract method stubs — exist to be overridden, not called

    if TYPE_CHECKING:     # always False at runtime — only runs for mypy/type checkers


# ── [html] controls the HTML report output ───────────────────────────────────
[html]

directory = htmlcov
# pytest --cov=app --cov-report=html → creates htmlcov/index.html
# open in browser: green = covered, red = not covered, line by line
# add htmlcov/ to .gitignore — generated file, not source code
```

---

### `# pragma: no cover` — Excluding Specific Lines

```python
# ✅ good use — only runs in production, impossible to test
if __name__ == '__main__':   # pragma: no cover
    uvicorn.run(app, host='0.0.0.0', port=8000)

# ❌ bad use — don't hide gaps in your test suite
def validate_payment(card):   # pragma: no cover  ← lazy, not acceptable
    ...
```

---

### Alternative — `pyproject.toml` (modern approach)

Skip `.coveragerc` and put everything in one file. Both do the same
thing — pick one, don't use both.

```toml
[tool.coverage.run]
source = ["app"]        # only measure 'app/' package
branch = true           # branch coverage, not just line coverage
omit = ["*/migrations/*", "app/__init__.py"]

[tool.coverage.report]
fail_under = 80         # CI fails automatically if coverage drops below 80%
exclude_lines = ["pragma: no cover", "raise NotImplementedError"]

[tool.coverage.html]
directory = "htmlcov"   # open htmlcov/index.html after --cov-report=html
```

> 💡 `fail_under = 80` is the most important production setting — makes
> coverage a hard gate on every commit rather than a number you check manually.

## 10. CI/CD — What Is It?

### The Problem Without CI

```
Developer A writes code → "works on my machine" → merges to main
Developer B pulls main  → tests fail on their machine
Production deploy       → bug ships to users
```

Everyone has a different local setup. CI replaces "works on my machine"
with **"works on a clean, neutral machine every single time."**

---

### CI — Continuous Integration

Every push or PR automatically triggers a fresh environment that runs
your full check suite:

```
Developer pushes code
        ↓
GitHub detects push → triggers workflow
        ↓
Runner (Ubuntu VM) spins up — clean slate, no local baggage
        ↓
Install deps → Run linter → Run tests → Check coverage
        ↓
All pass? → ✅ PR can be merged
Any fail? → ❌ PR is blocked — must fix before merge
```

Nobody merges broken code because the gate is automated — not a
manual review step someone can forget.

---

### CD — Continuous Delivery / Deployment

The natural next step after CI passes:

```
CI passes ✅
    ↓
Auto-deploy to staging   ← Continuous Delivery   (human approves prod)
    ↓
Auto-deploy to prod      ← Continuous Deployment (fully automated)
```

---

### What CI Runs — Typical Check Suite

```bash
pip install -r requirements.txt   # install dependencies fresh

ruff check .                       # linting   — catches style issues, unused imports
mypy app/                          # type check — catches type mismatches before runtime
pytest --cov=app --cov-fail-under=80   # tests + coverage gate
```

Each step must pass before the next runs — a linting failure stops
tests from even starting.

---

### Why CI Matters

```
"works on my machine"     → ❌ not a standard — everyone's machine differs
CI on a clean Ubuntu VM   → ✅ same environment every run, no exceptions

catch regressions         → immediately on push, not weeks later in prod
enforce quality           → coverage, linting, types — automated, not optional
protect main branch       → PRs cannot merge until every check is green
```

---

> 💡 CI is not just for big teams — even solo projects benefit. It
> forces you to keep tests passing and gives you confidence that the
> code on `main` always works.

## 11. GitHub Actions — Core Concepts

GitHub Actions is the CI/CD platform built into GitHub. Files go in `.github/workflows/`.

| Concept | What it is | Example |
|---------|------------|----------|
| **Workflow** | A YAML file defining automation | `ci.yml`, `deploy.yml` |
| **Trigger** | When does the workflow run? | `on: push`, `on: pull_request` |
| **Job** | A group of steps on one runner | `test`, `lint`, `deploy` |
| **Step** | One action or shell command | `- run: pytest` |
| **Runner** | The machine that runs the job | `ubuntu-latest`, `macos-latest` |
| **Action** | A reusable step (from GitHub marketplace) | `actions/checkout@v4`, `actions/setup-python@v5` |
| **Matrix** | Run same job with different inputs | Python 3.10, 3.11, 3.12 |

**File location:** `.github/workflows/ci.yml` (relative to repo root)

**Free tier:** 2,000 minutes/month for public repos (essentially unlimited for open source). For private repos, 2,000 minutes/month free.

## 11. GitHub Actions — Core Concepts

GitHub Actions is the CI/CD platform built directly into GitHub.
All automation is defined in YAML files inside `.github/workflows/`.

| Concept | What it is | Example |
|---------|------------|---------|
| **Workflow** | A YAML file defining automation | `ci.yml`, `deploy.yml` |
| **Trigger** | When does the workflow run? | `on: push`, `on: pull_request` |
| **Job** | A group of steps on one runner | `test`, `lint`, `deploy` |
| **Step** | One action or shell command | `- run: pytest` |
| **Runner** | The VM that runs the job | `ubuntu-latest`, `macos-latest` |
| **Action** | A reusable step from the marketplace | `actions/checkout@v4` |
| **Matrix** | Run same job with different inputs | Python 3.10, 3.11, 3.12 |

Think of it as: **Workflow contains Jobs, Jobs contain Steps, Steps run on a Runner.**

```
Workflow (ci.yml)
  └── Job: test (runs on ubuntu-latest)
        ├── Step: checkout code
        ├── Step: install python
        ├── Step: install deps
        └── Step: run pytest
```

> 💡 Free tier: 2,000 min/month for private repos. Public repos get
> essentially unlimited minutes — open source CI is free.

---

## 12. GitHub Actions Workflow — Anatomy

```yaml
# .github/workflows/ci.yml

name: CI                          # label shown in GitHub Actions UI

# ── TRIGGERS — when does this workflow run? ───────────────────────────────────
on:
  push:
    branches: [main, develop]     # run on every push to main or develop
  pull_request:
    branches: [main]              # run on every PR that targets main


# ── JOBS — what work gets done? ───────────────────────────────────────────────
jobs:
  test:                           # job name — arbitrary, shows in GitHub UI
    runs-on: ubuntu-latest        # spin up a fresh Ubuntu VM for this job

    # MATRIX — runs this entire job once per python version (3 parallel runs)
    # great for confirming your code works across multiple Python versions
    strategy:
      matrix:
        python-version: ['3.10', '3.11', '3.12']

    # ── STEPS — sequential commands inside the job ────────────────────────────
    steps:

      - name: Checkout code
        uses: actions/checkout@v4         # marketplace action: clones your repo
                                          # onto the runner — always the first step

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v5     # marketplace action: installs Python
        with:
          python-version: ${{ matrix.python-version }}   # injected from matrix above
          cache: 'pip'                    # caches pip dependencies — faster builds

      - name: Install dependencies
        run: pip install -r requirements.txt    # plain shell command — same as your terminal

      - name: Lint with ruff
        run: ruff check .                 # fails the job if linting errors found
                                          # job stops here — tests won't run on bad code

      - name: Run tests with coverage
        run: pytest --cov=app --cov-report=xml --cov-fail-under=80
        # --cov-report=xml   → writes coverage.xml for upload to codecov
        # --cov-fail-under=80 → job FAILS if coverage drops below 80%

      - name: Upload coverage report
        uses: codecov/codecov-action@v4   # optional: sends coverage.xml to codecov.io
        with:                             # codecov adds a coverage badge and PR comment
          file: coverage.xml
```

---

### How the Matrix Works

```
matrix: python-version: ['3.10', '3.11', '3.12']

→ GitHub runs THREE parallel jobs:
    test (3.10) ──┐
    test (3.11) ──┼── all run simultaneously
    test (3.12) ──┘

ALL three must pass for the PR to be mergeable
```

---

### Step Failure Stops the Job

```
checkout   ✅
setup python ✅
install deps ✅
ruff check   ❌ ← lint fails here — pytest never runs
pytest       ⏭ skipped
```

> 💡 The file lives at `.github/workflows/ci.yml` relative to your
> repo root — GitHub detects and runs it automatically. No extra setup
> needed beyond pushing the file.

## 13. Matrix Builds — Testing Multiple Python Versions

```yaml
strategy:
  matrix:
    python-version: ['3.10', '3.11', '3.12']
```

Runs the **entire job 3 times in parallel** — once per Python version.
If your code works on 3.11 but breaks on 3.10 due to a syntax
difference, the matrix catches it immediately.

---

### Matrix on Multiple Axes — OS × Python Version

```yaml
strategy:
  matrix:
    os: [ubuntu-latest, macos-latest, windows-latest]
    python-version: ['3.11', '3.12']
# creates 6 combinations (3 OS × 2 versions) — all run in parallel
```

```
ubuntu  + 3.11  ──┐
ubuntu  + 3.12  ──┤
macos   + 3.11  ──┼── all 6 run simultaneously
macos   + 3.12  ──┤
windows + 3.11  ──┤
windows + 3.12  ──┘
ALL must pass for PR to be mergeable
```

---

### Caching Dependencies — Important for Speed

```yaml
- uses: actions/setup-python@v5
  with:
    python-version: '3.11'
    cache: 'pip'     # caches ~/.cache/pip between runs — saves ~30s per run
```

```
without cache → reinstalls ALL packages from scratch every run  (~60s)
with cache    → reuses cached packages if requirements.txt unchanged (~5s)
```

GitHub invalidates the cache automatically when `requirements.txt`
changes — so you always get fresh packages when dependencies update.

---

## 14. Badges — Signal CI Status in README

Add these to your `README.md` to show live build and coverage status:

```markdown
![CI](https://github.com/username/repo/actions/workflows/ci.yml/badge.svg)
[![Coverage](https://codecov.io/gh/username/repo/branch/main/graph/badge.svg)](https://codecov.io/gh/username/repo)
```

These are **live images** — they update automatically on every push:

```
green  → passing   — all checks passed
red    → failing   — something broke
grey   → unknown   — workflow hasn't run yet
```

---

### Other Common Badges

```markdown
![Python](https://img.shields.io/badge/python-3.11-blue)
![License](https://img.shields.io/badge/license-MIT-green)
![PyPI](https://img.shields.io/pypi/v/your-package)   ← live version from PyPI
```

> 💡 Badges are not just cosmetic — a red CI badge on a PR is an
> immediate signal to reviewers that something needs fixing before merge.

---
## Summary — Cheat Sheet

### Integration Testing with FastAPI
```python
from fastapi.testclient import TestClient
from app.main import app

client = TestClient(app)

# GET
res = client.get('/users/1')
assert res.status_code == 200
assert res.json()['name'] == 'Alice'

# POST with body
res = client.post('/users', json={'name': 'Bob', 'email': 'b@b.com'})
assert res.status_code == 201

# With headers (e.g., auth token)
res = client.get('/protected', headers={'Authorization': 'Bearer token123'})

# 422 Validation error
res = client.post('/users', json={})   # missing required fields
assert res.status_code == 422
```

### Dependency Override
```python
@pytest.fixture
def client():
    app.dependency_overrides[get_db] = lambda: test_db
    yield TestClient(app)
    app.dependency_overrides.clear()    # always restore!
```

### Coverage
```bash
pytest --cov=app --cov-report=term-missing tests/
pytest --cov=app --cov-report=html tests/        # generates htmlcov/
pytest --cov=app --cov-fail-under=80 tests/      # fail if < 80%
```

### GitHub Actions — Key YAML
```yaml
name: CI
on: [push, pull_request]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: '3.11'
          cache: 'pip'
      - run: pip install -r requirements.txt
      - run: ruff check .
      - run: pytest --cov=app --cov-fail-under=80
```

### The Complete Testing Checklist
- [ ] Unit tests for all pure functions (no DB/HTTP)
- [ ] Integration tests for all API endpoints (GET, POST, PUT, DELETE)
- [ ] Test happy paths AND error cases (404, 422, 400, 403)
- [ ] Mock all external HTTP calls in unit tests
- [ ] Use dependency override for DB in integration tests
- [ ] `session.rollback()` to isolate tests from each other
- [ ] Coverage ≥ 80% (measure with `--cov-fail-under=80`)
- [ ] CI workflow runs on every push and PR
- [ ] CI blocks merge if tests fail